### ISSUE TITLE
Same Pad_Head value in ORT for SAME_UPPER/SAME_LOWER if get negative odd pad value

### ISSUE THREAD

**Issue by RunnerZhong** [microsoft/onnxruntime#10086]
**Describe the bug**
Implementation code in ORT is as below:
`case AutoPadType::SAME_LOWER: {
          int64_t legacy_target_size = (in_size + stride - 1) / stride;
          int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
          *pad_head = (pad_needed + 1) / 2;
          *pad_tail = pad_needed - *pad_head;
          *out_size = ComputeOutputSize(in_size, stride, kernel, pad_needed, dilation);
          break;
        }
        case AutoPadType::SAME_UPPER: {
          int64_t legacy_target_size = (in_size + stride - 1) / stride;
          int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
          *pad_head = pad_needed / 2;
          *pad_tail = pad_needed - *pad_head;
          *out_size = ComputeOutputSize(in_size, stride, kernel, pad_needed, dilation);
          break;
        }`

For below case, Input_shape = [1,3,500, 224], kernel=[5, 6], stride=[10,2], auto_pad = same_upper/lower
According to ONNX op SPEC,

> SAME_UPPER or SAME_LOWER: output_spatial_shape[i] = ceil(input_spatial_shape[i] / strides_spatial_shape[i])

> pad_shape[i] = (output_spatial_shape[i] - 1) * strides_spatial_shape[i] + kernel_spatial_shape[i] - input_spatial_shape[i]

we got pad_h_value = -5.

So, 
*pad_head = (pad_needed + 1) / 2;    # pad_head = -2
  
*pad_head = pad_needed / 2;   # pad_head = -2 too

**Urgency**
If there are particular important use cases blocked by this or strict project-related timelines, please share more information and dates. If there are no hard deadlines, please specify none.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):
- ONNX Runtime installed from (source or binary):
- ONNX Runtime version:
- Python version:
- Visual Studio version (if applicable):
- GCC/Compiler version (if compiling from source):
- CUDA/cuDNN version:
- GPU model and memory:

**To Reproduce**
- Describe steps/code to reproduce the behavior.
- Attach the ONNX model to the issue (where applicable) to expedite investigation.

**Expected behavior**

**Screenshots**

**Additional context**
Add any other context about the problem here. If the issue is about a particular model, please share the model details as well to facilitate debugging.

**Comment by skottmckay** [comment]
Is the issue that SAME_LOWER needs to take into account whether pad_needed is negative? Would something like the below fix it?

```cpp
        case AutoPadType::SAME_LOWER: {
          int64_t legacy_target_size = (in_size + stride - 1) / stride;
          int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
          *pad_head = (pad_needed < 0 ? pad_needed - 1 : pad_needed + 1) / 2;
          *pad_tail = pad_needed - *pad_head;
          *out_size = ComputeOutputSize(in_size, stride, kernel, pad_needed, dilation);
          break;
        }
```

Does this negative padding actually occur in a real model or is it more of a theoretical issue?

**Comment by gwang-msft** [comment]
This is probably an invalid operator, based on ONNX spec for pads of Conv/Pool ops, 
> pads : list of ints
> Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0.`

Or, at least the behavior is undefined for negative padding. (crop the input image?)

However, it will be nice ORT can throw on this case instead of moving forward with model execution.

**Comment by RunnerZhong** [comment]
> This is probably an invalid operator, based on ONNX spec for pads of Conv/Pool ops,
> 
> > pads : list of ints
> > Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0.`
> 
> Or, at least the behavior is undefined for negative padding. (crop the input image?)
> 
> However, it will be nice ORT can throw on this case instead of moving forward with model execution.

Yeah, Your case is only for explicit padding, but for auto_padding, the pad_value is calculated, it maybe negative (crop the input)

**Comment by RunnerZhong** [comment]
> Is the issue that SAME_LOWER needs to take into account whether pad_needed is negative? Would something like the below fix it?
> 
> ```c++
>         case AutoPadType::SAME_LOWER: {
>           int64_t legacy_target_size = (in_size + stride - 1) / stride;
>           int64_t pad_needed = (legacy_target_size - 1) * stride + kernel - in_size;
>           *pad_head = (pad_needed < 0 ? pad_needed - 1 : pad_needed + 1) / 2;
>           *pad_tail = pad_needed - *pad_head;
>           *out_size = ComputeOutputSize(in_size, stride, kernel, pad_needed, dilation);
>           break;
>         }
> ```
> 
> Does this negative padding actually occur in a real model or is it more of a theoretical issue?

It's not a REAL model, just be found during my own test. And for your demo fix, I think it's ok

**Comment by gwang-msft** [comment]
> > This is probably an invalid operator, based on ONNX spec for pads of Conv/Pool ops,
> > > pads : list of ints
> > > Padding for the beginning and ending along each spatial axis, it can take any value greater than or equal to 0.`
> > 
> > 
> > Or, at least the behavior is undefined for negative padding. (crop the input image?)
> > However, it will be nice ORT can throw on this case instead of moving forward with model execution.
> 
> Yeah, Your case is only for explicit padding, but for auto_padding, the pad_value is calculated, it maybe negative (crop the input)

It does not matter whether it is explicit padding or auto padding, eventually the auto padding will be converted to explicit padding in the model execution.

**Comment by bugxch** [comment]
I come across the same issue for the negative pads with onnx runtime. I also compared the implementation of onnx and tensorflow, 

- tensorflow, treated the negativa pads as zero, i.e.,  the total_pad = max(total_pad, 0), so it equals VALID mode in tensorflow
- onnx, crop the input with negative pads

I think it's an onnx bug because they are totally different implementation, and only one is correct for the common conv2d operator
